abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+`proxy_on_context_finalize`.
+
+Returns `Ok` on success.
+
+
+## Configuration
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_vm_start`
+
+* params:
+  - `i32 (uint32_t) vm_id`
+  - `i32 (size_t) vm_configuration_size`
+* returns:
+  - `i32 (bool) is_success`
+
+Called when the host environment starts WebAssembly Virtual Machine (WasmVm).
+
+Its configuration (of size `vm_configuration_size`) can be retrieved using
+`proxy_get_buffer`.
+
+Return value (`is_success`) indicates that configuration was processed correctly
+and that WasmVm started successfully.
+
+#### `proxy_on_plugin_start`
+
+* params:
+  - `i32 (uint32_t) plugin_id`
+  - `i32 (size_t) plugin_configuration_size`
+* returns:
+  - `i32 (bool) is_success`
+
+Called when the host environment starts plugin.
+
+Its configuration (of size `plugin_configuration_size`) can be retrieved using
+`proxy_get_buffer`.
+
+Return value (`is_success`) indicates that configuration was processed correctly
+and that plugin started successfully.
+
+
+## TCP/UDP/QUIC stream (L4) extensions
+
+Note: `downstream` means the connection between client and proxy, and `upstream`
+       means the connection between proxy and backend (aka “next hop”).
+
+Note: The same unique context identifier (`stream_id`) is shared by a pair of
+      `downstream` and `upstream` connections.
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_new_connection`
+
+* params:
+  - `i32 (uint32_t) stream_id`
+* returns:
+  - `i32 (proxy_action_t) next_action`
+
+Called on a new connection (`stream_id`).
+
+Return value (`next_action`) instructs the host environment what action to take:
+- `Continue` means that this connection should be accepted.
+- `Done` means that this connection should be accepted and that this extension
+  shouldn't be called again for the current stream.
+- `Pause` means that further processing should be paused until
+  `proxy_resume_stream` is called.
+- `Close` means that this connection should be rejected.
+
+#### `proxy_on_downstream_data`
+
+* params:
+  - `i32 (uint32_t) stream_id`
+  - `i32 (size_t) data_size`
+  - `i32 (bool) end_of_stream`
+* returns:
+  - `i32 (proxy_action_t) next_action`
+
+Called for each data chunk received from downstream (`stream_id`).
+
+Data (of size `data_size`) can be retrieved using `proxy_get_buffer`.
+
+Return value (`next_action`) instructs the host environment what action to take:
+- `Continue` means that data chunk should be forwarded upstream.
+- `EndStream` means that data chunk should be forwarded upstream and that this
+  connection should be closed afterwards.
+- `Done` means that data chunk should be forwarded upstream and that this
+  extension shouldn't be called again for the current stream.
+- `Pause` means that further processing should be paused until
+  `proxy_resume_stream` is called.
+- `WaitForMoreData` means that further processing should be paused and that host
```

Comment:
Is there a way for a network filter to short-circuit processing by writing some data back downstream, like "send_local_response" for HTTP filters? Imagine a filter that reads some authorization information from a connection and, if invalid, sends back an error message.

Reply:
We don't have that right now. I can add `proxy_send_bytes_downstream` if you want.